### PR TITLE
Service Route Peering initial commit

### PR DESCRIPTION
**Pull request by mmudigon**
This PR contains the following:

1. Service Route Peering backend module
2. Unit test cases
3. Integration test cases
4. prepare test case to create/delete VRFs,Service Nodes etc.
5. Fix to ethernet test case in dcnm_intf module. This is required to fix an issue which is exposed by changes to dcnm.py in the current PR

### CODE REVIEW

File: plugins/modules/dcnm_service_route_peering.py [hunk 1]
```diff
@@ -0,0 +1,3707 @@
+#!/usr/bin/python
+#
+# Copyright (c) 2020 Cisco and/or its affiliates.
```
Comment (mikewiebe) [inline plugins/modules/dcnm_service_route_peering.py:3]: Copyright should be `2021` for new files.

File: plugins/modules/dcnm_service_route_peering.py [hunk 2]
```diff
@@ -0,0 +1,3707 @@
...
+          - Name of the service node where the route peering is to be deployed
+        type: str
+        required: true
+      deploy_mode:
+        description:
+          - Type of service node. 
+        type: str
+        required: true
+        choices: ['IntraTenantFW', 'InterTenantFW', 'OneArmADC', 'TwoArmADC']
+      peering_option:
```
[the remark this answers is not in the document]
Comment (mikewiebe) [reply plugins/modules/dcnm_service_route_peering.py:84]: Maybe in `dcnm_srp_validate_and_build_srp_info`

File: plugins/modules/dcnm_service_route_peering.py [hunk 3]
```diff
@@ -0,0 +1,3707 @@
...
+            if self.module.params["state"] == "deleted":
+                # config for delete state is different. So validate deleted state differently
+                self.dcnm_srp_validate_delete_state_input(cfg)
+            elif self.module.params["state"] == "query":
+                # config for query state is different. So validate query state differently
+                self.dcnm_srp_validate_query_state_input(cfg)
+            # For 'overridden' state, we can have full config for a peering or just service node name alone.
+            # In the formar case go down to 'else' block to validate the full config
+            elif (self.module.params["state"] == "overridden") and (
+                item.get("name", None) is None
```
Comment (mikewiebe) [inline plugins/modules/dcnm_service_route_peering.py:1324]: I believe `item.get` will automatically return `None` if there is not value so not sure you need to default to `None`

**Review by mikewiebe — CHANGES_REQUESTED** [review]
Lot of good work here! I am not finished with the review but here are the first round of comments.

File: plugins/modules/dcnm_service_route_peering.py (hunk 1 above)
Comment (mmudigon) [reply plugins/modules/dcnm_service_route_peering.py:3]: Done

File: plugins/httpapi/dcnm.py [hunk 4]
```diff
@@ -78,12 +79,13 @@ def logout(self):
         path = '/rest/logout'
 
         try:
-            response, response_data = self.connection.send(path, {}, method=method, headers=self.headers, force_basic_auth=True)
+            response, response_data = self.connection.send(path, self.connection._auth['Dcnm-Token'], method=method, headers=self.headers, force_basic_auth=True)
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/httpapi/dcnm.py:82]: Yes, I have run other modules also

File: plugins/module_utils/network/dcnm/dcnm.py [hunk 5]
```diff
@@ -68,20 +68,36 @@ def validate_list_of_dicts(param_list, spec, module=None):
...
-                    if address.count('.') != 3:
-                        invalid_params.append('{} : Invalid IPv4 address syntax'.format(item))
+                        try:
+                            socket.inet_pton(socket.AF_INET, address)
+                        except socket.error:
+                            invalid_params.append('{} : Invalid IPv4 address syntax'.format(item))
+                        if address.count('.') != 3:
+                            invalid_params.append('{} : Invalid IPv4 address syntax'.format(item))
+
+                elif ((type == 'ipv6_subnet') or (type == 'ipv6')):
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/module_utils/network/dcnm/dcnm.py:87]: Done

File: plugins/module_utils/network/dcnm/dcnm.py [hunk 6]
```diff
@@ -228,49 +244,6 @@ def dcnm_get_ip_addr_info(module, sw_elem, ip_sn, hn_sn):
             msg_dict['Error'] = msg.format(sw_elem)
             raise module.fail_json(msg=json.dumps(msg_dict))
 
-# def dcnm_get_ip_addr_info(sw_elem, ip_sn):
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/module_utils/network/dcnm/dcnm.py:231]: Yes, the commented one is no longer required. so removing it

File: plugins/module_utils/network/dcnm/dcnm.py [hunk 7]
```diff
@@ -280,3 +253,10 @@ def dcnm_send(module, method, path, data=None, data_type='json'):
         return conn.send_request(method, path, data)
     elif (data_type == 'text'):
         return conn.send_txt_request(method, path, data)
+
+def dcnm_reset_connection(module):
+
+    conn = Connection(module._socket_path)
+
+    conn.logout()
+    return conn.login (conn.get_option("remote_user"), conn.get_option("password"))
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/module_utils/network/dcnm/dcnm.py:262]: conn.login() return s appropriate error. The modules can  check that directly.

File: plugins/modules/dcnm_service_route_peering.py [hunk 8]
```diff
@@ -0,0 +1,3707 @@
...
+      - merged
+      - replaced
+      - overridden
+      - deleted
+      - query
+    default: merged
+  attach:
+    description:
+      - A flag spcifying if the given route peering is to be attached to the specified service node
+    type: bool
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/modules/dcnm_service_route_peering.py:53]: Correcrted

[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/modules/dcnm_service_route_peering.py:53]: Used 'bool' in all places

File: plugins/modules/dcnm_service_route_peering.py [hunk 9]
```diff
@@ -0,0 +1,3707 @@
...
+
+Query:
+  Returns the current DCNM state for the route peerings listed in the playbook.
+
+CREATING ROUTE PEERINGS
+=======================
+
+INTRA-TENANT FIREWALL
+
+- name: Create different non-existing service route peering including all objects
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/modules/dcnm_service_route_peering.py:626]: Corrected

File: plugins/modules/dcnm_service_route_peering.py [hunk 10]
```diff
@@ -0,0 +1,3707 @@
...
+              profile:
+                ipv4_gw: 192.165.2.1/24                      # mandatory
+                ipv6_gw: 2005:db02::1/64                     # optional, default is ''
+                vlan_name: rp5-sn2-second-arm                # optional, default is ''
+                int_descr: "RP5 SN2 second arm intf"         # optional, default is ''
+                tag: 51112                                   # optional, default is 12345
+            rev_next_hop: 192.165.1.100                      # mandatory
+    attach: true
+    deploy: true
+    state: merged
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/modules/dcnm_service_route_peering.py:840]: Done

File: plugins/modules/dcnm_service_route_peering.py [hunk 11]
```diff
@@ -0,0 +1,3707 @@
...
+    state: deleted
+
+DELETE ROUTE PEERINGS
+=====================
+
+- name: Delete route peerings
+  cisco.dcnm.dcnm_service_route_peering:                    
+    fabric: test-fabric
+    service_fabric: external
+    config:
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/modules/dcnm_service_route_peering.py:925]: No. We need service node which is mandatory as per PRD. So config will be required. However the route peering name is optional.

File: plugins/modules/dcnm_service_route_peering.py [hunk 12]
```diff
@@ -0,0 +1,3707 @@
...
+        self.result = dict(changed=False, diff=[], response=[])
+
+    def log_msg(self, msg):
+
+        if self.fd == None:
+            self.fd = open("srp.log", "w+")
+        if self.fd != None:
+            self.fd.write(msg)
+
+    #############################################################################################################################
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/modules/dcnm_service_route_peering.py:1110]: Done. Updated all headers.

File: plugins/modules/dcnm_service_route_peering.py (hunk 3 above)
Comment (mmudigon) [reply plugins/modules/dcnm_service_route_peering.py:1324]: Some places I use a different value than None. So to make it consistent I use this style

File: plugins/modules/dcnm_service_route_peering.py [hunk 13]
```diff
@@ -0,0 +1,3707 @@
...
+    default: true
+  config:
+    description: 
+      - A list of dictionaries containing route peering and switch information
+    type: list
+    elements: dict
+    suboptions:
+      name:
+        description:
+          - Name of the route peering
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/modules/dcnm_service_route_peering.py:70]: Changed it to "A unique name which identifies the route peering"

File: plugins/modules/dcnm_service_route_peering.py [hunk 14]
```diff
@@ -0,0 +1,3707 @@
...
+        description:
+          - Name of the service node where the route peering is to be deployed
+        type: str
+        required: true
+      deploy_mode:
+        description:
+          - Type of service node. 
+        type: str
+        required: true
+        choices: ['IntraTenantFW', 'InterTenantFW', 'OneArmADC', 'TwoArmADC']
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/modules/dcnm_service_route_peering.py:83]: Updated the code to use snake_case and converted the same inside the module

File: plugins/modules/dcnm_service_route_peering.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/modules/dcnm_service_route_peering.py:84]: There is a check to default it out to 'static' if nothing is given. Also the module takes care of filling it only when required

File: plugins/modules/dcnm_service_route_peering.py [hunk 15]
```diff
@@ -0,0 +1,3707 @@
...
+                    elements: ipv4
+                    required: True
+              ipv4_neighobor:
+                description:
+                  - IPv4 neighbor address
+                  - NOTE: This object is applicable only when 'peering_option' is 'ebgp'
+                type: ipv4
+                required: True
+              ipv4_lo:
+                  - IPv4 loopback address
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/modules/dcnm_service_route_peering.py:192]: Added

File: plugins/modules/dcnm_service_route_peering.py [hunk 16]
```diff
@@ -0,0 +1,3707 @@
...
+                required: True
+                default: 12345
+              neigh_int_descr:
+                  - Description of the interface
+                  - NOTE: This object is applicable only when 'peering_option' is 'ebgp'
+                type: str
+                required: False
+                default: ''
+              local_asn:
+                  - Local ASN number
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/modules/dcnm_service_route_peering.py:375]: Added

File: plugins/modules/dcnm_service_route_peering.py [hunk 17]
```diff
@@ -0,0 +1,3707 @@
...
+
+EXAMPLES = """
+
+States:
+This module supports the following states:
+
+Merged:
+  Route Peerings defined in the playbook will be merged into the target fabric.
+
+  The route peerings listed in the playbook will be created if not already present on the DCNM
```
[the remark this answers is not in the document]
Comment (mmudigon) [reply plugins/modules/dcnm_service_route_peering.py:591]: Updated

**Pull request merged**